File: C++/3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int>> ret;
        sort(num.begin(), num.end());
        
        for (int i=0; i<num.size(); ++i) {
            int start = i+1, end = num.size()-1;
            while (start < end) {
                if (num[start] + num[end] == -num[i]) {
                    vector<int> triple = {num[i], num[start], num[end]};
                    ret.push_back(triple);
                    start++;
                    end--;
                    // To avoid non-unique answers.
                    while (start < end && num[start] == num[start-1]) start++;
                    while (start < end && num[end] == num[end+1]) end--;
                } else if (num[start] + num[end] > -num[i]) {
                    end--;
                } else {
                    start++;
                }
            }
            
            // To avoid non-unique answers.
            while (i<num.size()-1 && num[i] == num[i+1])
                i++;
        }
        
        return ret;
    }
};

File: javascript/3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  nums.sort((a, b) => (a-b)) // Sort by number

  const res = []
  for (let i = 0; i < nums.length - 2; ++i) {
      if (i > 0 && nums[i] === nums[i-1]) continue

      // Calculate sum and adjust j and k accordingly
      let j = i + 1, k = nums.length - 1
      while (j < k) {
          const sum = nums[i] + nums[j] + nums[k]
          if (sum === 0) {
              res.push([nums[i], nums[j], nums[k]])

              // To find unique answers,
              // we have to find the first different numbers other than original nums[j] and nums[k]
              // -2 -1 -1 1 1 2
              //    j       k
              //       j  k
              //    ^^^^^^^^^ these j, k create repeated tuples
              ++j
              while (nums[j] === nums[j-1]) ++j

              --k
              while (nums[k] === nums[k+1]) --k
          } else if (sum < 0) {
              ++j
          } else {
              --k
          }
      }
  }

  return res
};